### DISCUSSION TITLE
Invalid redirection on branches with `#` in their name

### DISCUSSION THREAD

**Discussion by k0ral** [community/community#85319] [Apps, API and Webhooks]
### Select Topic Area

Bug

### Body

I've created [a repository](https://github.com/k0ral/test-branch-redirection/branches) to showcase the issue, where:
1. I created branch `branch-name`
2. I renamed branch `branch-name` into `branch-name#42`
 
## Scenario
Send HTTP GET to https://api.github.com/repos/k0ral/test-branch-redirection/branches/branch-name and follow redirections until you get a valid HTTP 200 response.

## Current behavior
1. Request to https://api.github.com/repos/k0ral/test-branch-redirection/branches/branch-name gets an HTTP 301 response with `Location` header set to https://api.github.com/repos/k0ral/test-branch-redirection/branches/branch-name#42 (notice that the `#` character isn't escaped)
2. Following redirection: request to https://api.github.com/repos/k0ral/test-branch-redirection/branches/branch-name#42 gets an HTTP 301 response with `Location` header set to https://api.github.com/repos/k0ral/test-branch-redirection/branches/branch-name#42 (i.e. itself)
3. ... Loop forever

## Desired behavior
1. Request to https://api.github.com/repos/k0ral/test-branch-redirection/branches/branch-name gets an HTTP 301 response with `Location` header set to https://api.github.com/repos/k0ral/test-branch-redirection/branches/branch-name%2342 **(`#` is escaped)**
2. Following redirection: request to https://api.github.com/repos/k0ral/test-branch-redirection/branches/branch-name%2342 gets an HTTP 200 response

## Analysis
The unescaped `#` character is interpreted as an URL anchor, ignored by the API, so https://api.github.com/repos/k0ral/test-branch-redirection/branches/branch-name#42 is equivalent to https://api.github.com/repos/k0ral/test-branch-redirection/branches/branch-name , hence the infinite redirection.

**Comment by Naimulnobel** [comment]
This seems to be an issue with how GitHub's API handles branch names that contain special characters like `#`. The `#` character is a reserved character in URLs and is typically used to indicate a fragment or anchor. When it's included in a branch name, it seems to cause the API to return a redirect URL that contains an unescaped `#` character, which leads to an infinite redirect loop.



1. **Avoid using special characters in branch names.** While it's technically possible to include special characters like `#` in branch names, it can lead to issues like this one. Using only alphanumeric characters, hyphens, and underscores in branch names can help avoid these issues.

2. **Manually escape the `#` character in the redirect URL.** If you're making requests to the GitHub API programmatically, you could add logic to your code to replace `#` with `%23` in the redirect URL before following the redirect.

You might also want to report this issue to GitHub. They may be able to fix the issue in their API, or they might have additional suggestions for workarounds.

**Comment by ghostinhershell** [comment]
Hi @k0ral, 

Thanks so much for reaching out in our community. It sounds like this may be expected behavior. 

When using a CLI, you might have situations where a branch or tag name contains special characters that have a special meaning for your shell environment. To use these characters safely in a Git command, they must be quoted or escaped, otherwise the command may have unintended effects: [Dealing with special characters in branch and tag names](https://docs.github.com/en/get-started/using-git/dealing-with-special-characters-in-branch-and-tag-names).
